pkg/dynamic/client_test.go

```diff
@@ -170,14 +171,27 @@ func TestListSourcesUsingGVKs(t *testing.T) {
 
 }
 
+var gvrToListKind = map[schema.GroupVersionResource]string{
+	{Group: "apiextensions.k8s.io", Version: "v1", Resource: "customresourcedefinitions"}: "CustomResourceDefinitionList",
+	{Group: "sources.knative.dev", Version: "v1", Resource: "apiserversources"}:           "ApiServerSourceList",
+	{Group: "sources.knative.dev", Version: "v1", Resource: "containersources"}:           "ContainerSourceList",
+	{Group: "sources.knative.dev", Version: "v1", Resource: "sinkbindings"}:               "SinkBindingList",
+	{Group: "sources.knative.dev", Version: "v1beta2", Resource: "pingsources"}:           "PingSourceList",
+	{Group: "sources.knative.dev", Version: "v1alpha1", Resource: "kafkasources"}:         "KafkaSourceList",
+	{Group: "eventing.knative.dev", Version: "v1", Resource: "brokers"}:                   "BrokerList",
+	{Group: "eventing.knative.dev", Version: "v1", Resource: "subscriptions"}:             "SubscriptionList",
+	{Group: "eventing.knative.dev", Version: "v1", Resource: "channels"}:                  "ChannelList",
+	{Group: "messaging.knative.dev", Version: "v1", Resource: "inmemorychannels"}:         "InMemoryChannelsList",
+}
+
 // createFakeKnDynamicClient gives you a dynamic client for testing containing the given objects.
 // See also the one in the fake package. Duplicated here to avoid a dependency loop.
 func createFakeKnDynamicClient(testNamespace string, objects ...runtime.Object) KnDynamicClient {
 	scheme := runtime.NewScheme()
```

Comment:
You could technically drop the need for the gvrToListKind if you add the correct types to the scheme

ie. for k8s types invoke `scheme.AddToScheme(scheme)` from https://github.com/kubernetes/client-go/blob/1bccfc8c60977e2ce3235394daff128996818f7b/kubernetes/scheme/register.go#L135

You can do this with eventing and serving and save yourself from maintaining this list. Unsure if there's something that prevents you from importing those.

Reply:
I also had some issues where I couldn't provide unstructured types to the fake dynamic client. There would be an error with the way the fake client interacts with its object tracker

So I wrote some code for the fake injection client to rewire these schemes to use unstructured.* types
https://github.com/knative/pkg/pull/2145/files#diff-8438a08a79f266761b674dbf8748084e62be792ce56e7d7b598d148c069946e5L40

Unsure if you would need to do the same - if so maybe we pull that bit of logic out out injection and into some sort of fakedynamic package in `knative.dev/pkg`